server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -585,18 +585,23 @@ public ParsedDocument index(IndexItem item) throws IOException {
                     columnConstraints
                 );
             }
+            undeterministicValues.clear();
             for (var entry : synthetics.entrySet()) {
                 ColumnIdent column = entry.getKey();
                 if (!column.isRoot()) {
                     continue;
                 }
                 Synthetic synthetic = entry.getValue();
-                ValueIndexer<Object> indexer = synthetic.indexer();
+
                 Object value = synthetic.input().value();
+                if (undeterministicIdents.contains(column)) {
```

Comment:
`ObjectIndexer#indexValue()#L115` is also calling: `innerValue = synthetic.input().value();` shouldn't we store and use this as well?
Can you please add test with returning a subcol non-deterministic value?

Reply:
Good point, let me check it

Reply:
That's a bug indeed, thanks for catching!

I pushed a test update (both unit/itest) will provide a fix.

Reply:
Could we do the other way around?
Can we store already all generated/default values in the ctor, and use them for indexing when required? Maybe we need to pass a different structure through `indexValue` but I'm not sure how this will work if actual index values are used in those expressions to calculate the final generated value. (Keep in mind the issue is the same both for `default` and `generated as....`.

Reply:
If we store gen/default columns in ctor we will get same values for such columns across the whole table which will be a bug.

We even had this issue and it got fixed in https://github.com/crate/crate/pull/12126/files#diff-6b4ff31a02425d333cc54423875ca7ab94432d3399bd0d6b45797763cbb5dcd7R86-R89

Reply:
I just don't like that in `getImplementation()` we already calculate those values:
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L190
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L196
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L222
and then during indexing we need to calculate again.

Reply:
> If we store gen/default columns in ctor we will get same values for such columns across the whole table which will be an issue.
> 
> We even had this issue and it got fixed in https://github.com/crate/crate/pull/12126/files#diff-6b4ff31a02425d333cc54423875ca7ab94432d3399bd0d6b45797763cbb5dcd7R86-R89
> 
> We have to compute synthetics on each `index()` but now we are also caching/reusing non-deterministic values.

Makes sense, thx!

Reply:
I pushed a fixup(s), could you take another look? `@Repeat` is temporal, I will revert it (or maybe let's drop Itest?)

Reply:
sorry, just saw another comment to keep itest, will only remove `repeat` then